File: CCC/CCC '03 S3 - Floor Plan.py
f=int(input())
r=int(input())
c=int(input())
rms=[]
d={}
fls=[]
for i in range(r):
   rms.append(input())
for i in range(r):
   for j in range(c):
      if rms[i][j]=="I": d[tuple([i, j])]=1
      else: d[tuple([i, j])]=0
def dfs(x, y, d):
   s=[]
   trs=0
   d[tuple([x, y])]=1
   s.append(tuple([x, y]))
   while len(s)>0:
      t=s.pop()
      trs+=1
      if t[0]>0 and d[tuple([t[0]-1, t[1]])]==0: 
         s.append(tuple([t[0]-1, t[1]]))
         d[tuple([t[0]-1, t[1]])]=1
      if t[0]<(len(rms)-1) and d[tuple([t[0]+1, t[1]])]==0: 
         s.append(tuple([t[0]+1, t[1]]))
         d[tuple([t[0]+1, t[1]])]=1
      if t[1]>0 and d[tuple([t[0], t[1]-1])]==0: 
         s.append(tuple([t[0], t[1]-1]))
         d[tuple([t[0], t[1]-1])]=1
      if t[1]<(len(rms[0])-1) and d[tuple([t[0], t[1]+1])]==0: 
         s.append(tuple([t[0], t[1]+1]))
         d[tuple([t[0], t[1]+1])]=1
   return tuple([trs, d]) 
for i in range(r):
   for j in range(c):
      if d[tuple([i, j])]==0:
         tmt=dfs(i, j, d)
         fls.append(tmt[0])
         d=tmt[1]
fls.sort()
fls.reverse()
i=0
while i<len(fls) and f-fls[i]>=0: 
   f-=fls[i]
   i+=1
if i==1:print(i,"room,", f, "square metre(s) left over")
else: print(i,"rooms,", f, "square metre(s) left over")

File: CCC/CCC '03 S3 - Floor Plan.cpp
#include <bits/stdc++.h>
using namespace std;
const int INF = 0x3f3f3f3f;
pair<int, int> dirs[]={{-1,0}, {1,0}, {0,-1}, {0,1}};
int r, c;
bool vld(int cr, int cc){
	return (cr<r&&cr>=0&&cc<c&&cc>=0);
}
string plan[30];
bool vstd[30][30];
int dfs(int x, int y){
	if(!vld(x, y) || vstd[x][y] || plan[x][y]!='.') return 0;
	int tt = 1;
	vstd[x][y]=1;
	for(pair<int, int> tp: dirs){
		tt+=dfs(x+tp.first, y+tp.second);
	}
	return tt;
}
vector<int> rms;
int main() {
	int fl;
	cin>>fl>>r>>c;
	for(int i=0; i<r; i++){
		cin>>plan[i];
	}
	for(int i=0; i<r; i++){
		for(int j=0; j<c; j++){
			if(plan[i][j]=='.' && !vstd[i][j]){
				rms.push_back(dfs(i,j));
			}
		}
	}
	sort(rms.begin(), rms.end());
	int tt = 0;
	for(int i=rms.size()-1; i>=0; i--){
		if(rms[i]<=fl){
			tt++; fl-=rms[i];
		}
		else{
			break;
		}
	}
	if (tt==1){
		printf("%i room, %i square metre(s) left over", tt, fl);
	}
	else{
		printf("%i rooms, %i square metre(s) left over", tt, fl);
	}
    return 0;
}